src/main/cpp/src/map_utils.cu

```diff
@@ -0,0 +1,633 @@
+                                             rmm::cuda_stream_view stream) {
+  if (input.is_empty()) {
+    return cudf::detail::make_device_uvector_async<char>(std::vector<char>{'[', ']'}, stream);
+  }
+
+  auto const d_strings = cudf::column_device_view::create(input, stream);
+  auto const transform_it = thrust::make_counting_iterator(0);
+  auto const output_size =
+      2 + // two extra bracket characters '[' and ']'
+      thrust::transform_reduce(
+          rmm::exec_policy(stream), transform_it, transform_it + input.size(),
+          [d_strings = *d_strings,
+           n_rows = input.size()] __device__(cudf::size_type idx) -> int64_t {
+            auto bytes = d_strings.is_null(idx) ?
+                             2 // replace null with "{}"
+                             :
+                             d_strings.element<cudf::string_view>(idx).size_bytes();
+            if (idx + 1 < n_rows) {
+              bytes += 1; // append `,` character to each input rows, except the last one
+            }
+            return static_cast<int64_t>(bytes);
+          },
+          int64_t{0}, thrust::plus{});
+
+  CUDF_EXPECTS(output_size <= static_cast<int64_t>(std::numeric_limits<cudf::size_type>::max()),
+               "The input json column is too large and causes overflow.");
+
+  auto const joined_input = cudf::strings::detail::join_strings(
+      cudf::strings_column_view{input},
+      cudf::string_scalar(","),  // append `,` character between the input rows
+      cudf::string_scalar("{}"), // replacement for null rows
+      stream, rmm::mr::get_current_device_resource());
+  auto const joined_input_child =
+      joined_input->child(cudf::strings_column_view::chars_column_index);
+  auto const joined_input_size_bytes = joined_input_child.size();
+  CUDF_EXPECTS(joined_input_size_bytes + 2 == output_size, "Incorrect output size computation.");
+
+  // We want to concatenate 3 strings: "[" + joined_input + "]".
+  // For efficiency, let's use memcpy instead of `cudf::strings::detail::concatenate`.
+  auto output = rmm::device_uvector<char>(joined_input_size_bytes + 2, stream);
+  CUDF_CUDA_TRY(cudaMemsetAsync(output.data(), static_cast<int>('['), 1, stream.value()));
+  CUDF_CUDA_TRY(cudaMemcpyAsync(output.data() + 1, joined_input_child.view().data<char>(),
+                                joined_input_size_bytes, cudaMemcpyDefault, stream.value()));
+  CUDF_CUDA_TRY(cudaMemsetAsync(output.data() + joined_input_size_bytes + 1, static_cast<int>(']'),
+                                1, stream.value()));
+
+#ifdef DEBUG_FROM_JSON
+  print_debug<char, char>(output, "Processed json string", "", stream);
+#endif
+  return output;
+}
+
+// Check and throw exception if there is any parsing error.
+void throw_if_error(rmm::device_uvector<PdaTokenT> const &tokens,
+                    rmm::device_uvector<SymbolOffsetT> const &token_indices,
+                    rmm::cuda_stream_view stream) {
+  auto const error_count =
```

Comment:
We don't use error count outside of knowing if there was an error, why not use the find below instead to reduce the number of operations? I know this is only in an error state and may not be worth it, so feel free to push back on this.

Reply:
The `thrust::find(+_if)`  APIs are known to have performance regression. We talked about that many times in cudf PRs. As such, it is called only when we will throw for sure.